src/internal/connector/onedrive/restore.go

```diff
@@ -36,17 +42,48 @@ func RestoreCollections(
 	var (
 		restoreMetrics support.CollectionMetrics
 		restoreErrors  error
+		metrics        support.CollectionMetrics
+		folderPerms    map[string][]UserPermission
+		canceled       bool
+
+		// permissionIDMappings is used to map between old and new id
+		// of permissions as we restore them
+		permissionIDMappings = map[string]string{}
 	)
 
 	errUpdater := func(id string, err error) {
 		restoreErrors = support.WrapAndAppend(id, err, restoreErrors)
 	}
 
+	// Reorder collections so that the parents directories are created
+	// before the child directories
+	sort.Slice(dcs, func(i, j int) bool {
+		return dcs[i].FullPath().String() < dcs[j].FullPath().String()
+	})
+
+	parentPermissions := map[string][]UserPermission{}
+
 	// Iterate through the data collections and restore the contents of each
 	for _, dc := range dcs {
-		temp, canceled := RestoreCollection(ctx, service, dc, OneDriveSource, dest.ContainerName, deets, errUpdater)
+		parentPerms, ok := parentPermissions[dc.FullPath().String()]
+		if !ok {
+			if len(dc.FullPath().Elements()) == 7 {
```

Comment:
nit: this feels kind of brittle since it assumes all the prefix directories for a path have a fixed number of elements. There are some functions for dealing with OneDrive paths that maybe we can leverage. The `path` package below is `src/pkg/path` no the stdlib one
```suggestion
			onedrivePath, err := path.ToOneDrivePath(dc.FullPath())
			if err != nil {
				// some error handling here
			}

			if len(onedrivePath.Folders) == 0 {
```